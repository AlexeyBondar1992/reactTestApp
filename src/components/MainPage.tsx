import React, {Component, ReactNode} from 'react';
import CardList from '../components/CardList';
import SearchBox from '../components/SearchBox';
import Scroll from '../components/Scroll';
import ErrorBoundary from '../components/ErrorBoundary';
import Header from '../components/Header';
import './MainPage.css';
import {IStateRobots, IStateSearch} from '../interfaces';

export interface IMainPageProps extends IMapSateToProps, IMapDispatchToProps {
    children?: ReactNode;
}

export interface IMapSateToProps extends IStateSearch, IStateRobots {
}

export interface IMapDispatchToProps {
    onSearchChange: (event: React.SyntheticEvent<HTMLInputElement>) => void;
    onRequestRobots: () => void;
}

class MainPage extends Component<IMainPageProps> {
    componentDidMount () {
        this.props.onRequestRobots();
    }

    filterRobots = () => {
        return this.props.robots
            .filter(robot => robot.name.toLowerCase().includes(this.props.searchField.toLowerCase()));
    };

    render () {
        const { onSearchChange, isPending } = this.props;
        if (isPending) {
            return <h1 className='tc'>Loading</h1>;
        }
        return (
            <div className='tc'>
                <Header />
                <SearchBox searchChange={onSearchChange} />
                <Scroll>
                    <ErrorBoundary>
                        <CardList robots={this.filterRobots()}/>
                    </ErrorBoundary>
                </Scroll>
            </div>
        );
    }
}

export default MainPage;
