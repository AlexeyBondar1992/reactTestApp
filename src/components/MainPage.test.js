import { shallow } from 'enzyme/build';
import toJson from 'enzyme-to-json';
import React from 'react';
import MainPage from './MainPage';

const mockProps = {
    onRequestRobots: jest.fn(),
    robots: [],
    searchField: '',
    isPending: false
};

let wrapper;

describe('MainPage', () => {
    beforeEach(() => {
        wrapper = shallow(<MainPage { ...mockProps } />);
    });

    it('should render MainPage without errors', () => {
        expect(toJson(wrapper)).toMatchSnapshot();
    });

    it('should filter robots correctly', () => {
        const props = {
            ...mockProps,
            searchField: 'j',
            robots: [{
                id: 3,
                name: 'John',
                email: 'john@gmail.com'
            }]
        };
        const wrapper2 = shallow(<MainPage { ...props } />);
        const instance = wrapper2.instance();

        expect(instance.filterRobots()).toEqual([props.robots[0]]);
    });
});
