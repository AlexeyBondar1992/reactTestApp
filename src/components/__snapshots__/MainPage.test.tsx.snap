// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MainPage should render MainPage without errors 1`] = `
<div
  className="tc"
>
  <Header />
  <SearchBox />
  <Scroll>
    <ErrorBoundary>
      <CardList
        robots={Array []}
      />
    </ErrorBoundary>
  </Scroll>
</div>
`;
